scalyr_agent/builtin_monitors/kubernetes_events_monitor.py

```diff
@@ -0,0 +1,478 @@
+
+            # check if this node is the current leader node
+        except Exception, e:
+            global_log.log( scalyr_logging.DEBUG_LEVEL_0, "Unexpected error checking for leader: %s" % (str(e)))
+
+        return leader is not None and self._node_name == leader
+
+    def run(self):
+        """Begins executing the monitor, writing metric output to logger.
+        """
+        try:
+            k8s = KubernetesApi()
+
+            # create the k8s cache
+            k8s_cache = KubernetesCache( k8s, self._logger,
+                cache_expiry_secs=self.__k8s_cache_expiry_secs,
+                cache_purge_secs=self.__k8s_cache_purge_secs,
+                namespaces_to_ignore=self.__k8s_namespaces_to_ignore
+                )
+
+            if self.__log_watcher:
+                self.log_config = self.__log_watcher.add_log_config( self, self.log_config )
+
+            pod_name = k8s.get_pod_name()
+            self._node_name = k8s.get_node_name( pod_name )
+            cluster_name = k8s.get_cluster_name()
+
+            last_event = None
+            last_resource = 0
+
+            last_check = time.time() - self._leader_check_interval
+
+            while not self._is_stopped():
+                current_time = time.time()
+
+                # if we are the leader, we could be going through this loop before the leader_check_interval
+                # has expired, so make sure to only check for a new leader if the interval has expired
+                if last_check + self._leader_check_interval <= current_time:
+                    last_check = current_time
+                    # check if we are the leader
+                    if not self._is_leader( k8s ):
+                        #if not, then sleep and try again
+                        global_log.log( scalyr_logging.DEBUG_LEVEL_0, "Leader is %s" % (str(self._current_leader)) )
+                        self._sleep_but_awaken_if_stopped(self._leader_check_interval)
+                        continue
+
+                    global_log.log( scalyr_logging.DEBUG_LEVEL_0, "Leader is %s" % (str(self._current_leader)) )
+                try:
+                    # start streaming events
+                    lines = k8s.stream_events( last_event=last_event )
+
+                    json = {}
+                    for line in lines:
+                        try:
+                            json = json_lib.parse(line)
+                        except Exception, e:
+                            global_log.warning( "Error parsing event json: %s, %s, %s" % (line, str(e), traceback.format_exc() ) )
+                            continue
+
+                        obj = json.get( "object", JsonObject() )
+
+                        # check to see if the resource version we are using has expired
+                        if json.get( 'type', '' ) == 'ERROR':
+                            if obj.get( 'kind', '' ) == 'Status' and obj.get( 'reason', '' ) == 'Expired':
```

Comment:
Right a method called `is_resource_expired` (or something like that) and use it here.

Reply:
Done